// @nodesecure/rc/dist/rc.js
// Import Node.js Dependencies
import os from "node:os";
import path from "node:path";
// Import Internal Dependencies
import { GLOBAL_CONFIGURATION_DIRECTORY } from "./constants.js";
import { loadJSONSchemaSync } from "./schema/loader.js";
import { generateCIConfiguration } from "./projects/ci.js";
import { generateReportConfiguration } from "./projects/report.js";
// CONSTANTS
export const JSONSchema = loadJSONSchemaSync();
/**
 * @example
 * generateDefaultRC("complete");
 * generateDefaultRC(["ci", "report"]); // minimal + ci + report
 */
export function generateDefaultRC(mode = "minimal") {
    const modes = new Set(typeof mode === "string" ? [mode] : mode);
    const minimalRC = {
        version: "1.0.0",
        i18n: "english",
        strategy: "npm"
    };
    const complete = modes.has("complete");
    return Object.assign(minimalRC, complete || modes.has("ci") ? generateCIConfiguration() : {}, complete || modes.has("report") ? generateReportConfiguration() : {});
}
/**
 * Dedicated directory for NodeSecure to store the configuration in the os HOME directory.
 */
export function homedir() {
    return path.join(os.homedir(), GLOBAL_CONFIGURATION_DIRECTORY);
}
export { generateCIConfiguration, generateReportConfiguration };
