// @nodesecure/ci/dist/src/configuration/standard/nsci.js
// Import Third-party Dependencies
import * as jsxray from "@nodesecure/js-x-ray";
import { IgnorePatterns } from "../external/nodesecure/ignore-file.js";
export const vulnStrategy = {
    npm: "NPM_AUDIT",
    node: "SECURITY_WG",
    snyk: "SNYK",
    sonatype: "SONATYPE",
    none: "NONE"
};
export const vulnSeverity = {
    MEDIUM: "medium",
    HIGH: "high",
    CRITICAL: "critical",
    ALL: "all"
};
export const warnings = {
    ERROR: "error",
    OFF: "off",
    WARNING: "warning"
};
export const warningNames = Object.keys(jsxray.warnings);
export const reporterTarget = {
    CONSOLE: "console",
    HTML: "html"
};
function generateDefaultRC() {
    return {
        rootDir: process.cwd(),
        strategy: vulnStrategy.npm,
        reporters: [reporterTarget.CONSOLE],
        vulnerabilitySeverity: vulnSeverity.MEDIUM,
        warnings: warnings.ERROR,
        ignorePatterns: IgnorePatterns.default()
    };
}
export const defaultNsciRuntimeConfiguration = generateDefaultRC();
