// @myunisoft/httpie/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mockErrors = exports.MockAgent = exports.File = exports.FormData = exports.Headers = exports.getGlobalDispatcher = exports.setGlobalDispatcher = exports.fetch = exports.ProxyAgent = exports.Agent = exports.DEFAULT_HEADER = exports.computeURI = exports.agents = exports.policies = void 0;
// Import Third-party Dependencies
const undici_1 = require("undici");
Object.defineProperty(exports, "Agent", { enumerable: true, get: function () { return undici_1.Agent; } });
Object.defineProperty(exports, "ProxyAgent", { enumerable: true, get: function () { return undici_1.ProxyAgent; } });
Object.defineProperty(exports, "fetch", { enumerable: true, get: function () { return undici_1.fetch; } });
Object.defineProperty(exports, "setGlobalDispatcher", { enumerable: true, get: function () { return undici_1.setGlobalDispatcher; } });
Object.defineProperty(exports, "getGlobalDispatcher", { enumerable: true, get: function () { return undici_1.getGlobalDispatcher; } });
Object.defineProperty(exports, "Headers", { enumerable: true, get: function () { return undici_1.Headers; } });
Object.defineProperty(exports, "FormData", { enumerable: true, get: function () { return undici_1.FormData; } });
Object.defineProperty(exports, "File", { enumerable: true, get: function () { return undici_1.File; } });
Object.defineProperty(exports, "MockAgent", { enumerable: true, get: function () { return undici_1.MockAgent; } });
Object.defineProperty(exports, "mockErrors", { enumerable: true, get: function () { return undici_1.mockErrors; } });
__exportStar(require("./request"), exports);
__exportStar(require("./stream"), exports);
__exportStar(require("./retry"), exports);
exports.policies = __importStar(require("./policies"));
var agents_1 = require("./agents");
Object.defineProperty(exports, "agents", { enumerable: true, get: function () { return agents_1.agents; } });
Object.defineProperty(exports, "computeURI", { enumerable: true, get: function () { return agents_1.computeURI; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "DEFAULT_HEADER", { enumerable: true, get: function () { return utils_1.DEFAULT_HEADER; } });
