// @nodesecure/ci/dist/src/lib/console-printer/index.js
import kleur from "kleur";
function createConsoleMessage(msg) {
    return {
        message: msg,
        bold() {
            this.message = consolePrinter.decoration.bold(this.message);
            return this;
        },
        italic() {
            this.message = consolePrinter.decoration.italic(this.message);
            return this;
        },
        underline() {
            this.message = consolePrinter.decoration.underline(this.message);
            return this;
        },
        print() {
            console.log(`\n ${this.message}`);
        }
    };
}
export const consolePrinter = {
    font: {
        standard: (message) => createConsoleMessage(kleur.white(message)),
        highlight: (message) => createConsoleMessage(kleur.magenta(message)),
        info: (message) => createConsoleMessage(kleur.yellow(message)),
        error: (message) => createConsoleMessage(kleur.red(message)),
        success: (message) => createConsoleMessage(kleur.bgGreen().white(message)),
        failure: (message) => createConsoleMessage(kleur.bgRed().white(message))
    },
    decoration: {
        underline: (message) => kleur.underline(message),
        italic: (message) => kleur.italic(message),
        bold: (message) => kleur.bold(message)
    },
    util: {
        concatOutputs: (messages, delimiter = " ") => createConsoleMessage(messages.join(delimiter))
    }
};
